src/lustre/lustreAstDependencies.ml

```diff
@@ -665,22 +670,44 @@ let rec vars_with_flattened_nodes: node_summary -> int -> LA.expr -> LA.SI.t
 (** get all the variables and flatten node calls using 
     the node summary for an expression *)
              
-let rec mk_contract_eqn_map: LA.contract_node_equation IMap.t -> LA.contract -> ((LA.contract_node_equation IMap.t), [> error]) result
-  = fun m ->
+(* The first map is a map from identifiers to contract equations, while the
+   second map only keeps track of the identifiers. We need both because
+   we may have multiple identifiers associated with the same equation, but
+   we still need to check for redeclaration. *)
```

Comment:
Although this solution works, I would prefer a solution where only one map and one check are used. I suggest to use a map from identifiers to optional equations, i.e. `(LA.contract_node_equation option) IMap.t`, so you pass `Some gc` to  `check_and_add` for all cases except for the elements in the tail of the list of a ghost variable declaration. Then, in `extract_decls` you only add a declaration if `find_opt` returns `Some (Some i')`.